graphql_compiler/schema/__init__.py

```diff
@@ -340,36 +340,47 @@ def _serialize_date(value: Any) -> str:
 
 def _parse_date_value(value: Any) -> date:
     """Deserialize a Date object from its proper ISO-8601 representation."""
-    return arrow.get(value, "YYYY-MM-DD").date()
+    if type(value) == date:
+        # We prefer exact type equality instead of isinstance() because datetime objects
+        # are subclasses of date but are not interchangeable for dates for our purposes.
+        return value
+    elif isinstance(value, str):
+        return arrow.get(value, "YYYY-MM-DD").date()
+    else:
+        raise ValueError(
+            f"Expected a date object or its ISO-8601 string representation. "
+            f"Got {value} of type {type(value)} instead."
+        )
 
 
 def _serialize_datetime(value: Any) -> str:
     """Serialize a DateTime object to its proper ISO-8601 representation."""
-    # Python datetime.datetime is a subclass of datetime.date, but in this case, the two are not
-    # interchangeable. Rather than using isinstance, we will therefore check for exact type
-    # equality.
-    #
     # We don't allow Arrow objects as input since it seems that Arrow objects are always tz aware.
     # This is supported by the fact that the `.naive` Arrow method returns a datetime object instead
     # of an Arrow object.
-    if type(value) == datetime and value.tzinfo is None:
+    if isinstance(value, datetime) and value.tzinfo is None:
         return value.isoformat()
     else:
         raise ValueError(
-            f"Expected a timezone naive datetime object. Got {value} of type {type(value)} instead."
+            f"Expected a timezone-naive datetime object. Got {value} of type {type(value)} instead."
         )
 
 
 def _parse_datetime_value(value: Any) -> datetime:
```

Comment:
```suggestion
def _parse_datetime_value(value: Union[datetime, str]) -> datetime:
```

Reply:
I think this is not allowed for the same reason as above — it does not follow the upstream contract, which is that `parse()` and `serialize()` on GraphQL type objects (which both this function and the function above are used as) take `Any` arguments.
https://github.com/graphql-python/graphql-core/blob/master/src/graphql/type/scalars.py#L105